sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSqlEnv.java

```diff
@@ -18,111 +18,64 @@
-import org.apache.calcite.tools.Frameworks;
 
 /**
  * {@link BeamSqlEnv} prepares the execution context for {@link BeamSql} and {@link BeamSqlCli}.
  *
- * <p>It contains a {@link SchemaPlus} which holds the metadata of tables/UDF functions, and
- * a {@link BeamQueryPlanner} which parse/validate/optimize/translate input SQL queries.
+ * <p>This class mainly provides methods to register {@code UDF} and {@code UDAF}, which can be
+ * used
  */
-public class BeamSqlEnv implements Serializable{
-  transient SchemaPlus schema;
-  transient BeamQueryPlanner planner;
-
-  public BeamSqlEnv() {
-    schema = Frameworks.createRootSchema(true);
-    planner = new BeamQueryPlanner(schema);
-  }
+public class BeamSqlEnv implements Serializable {
+  private Map<String, Class<? extends BeamSqlUdf>> udfs = new HashMap<>();
+  private Map<String, Class<? extends BeamSqlUdaf>> udafs = new HashMap<>();
+  private Map<String, SerializableFunction> sfnUdfs = new HashMap<>();
 
   /**
    * Register a UDF function which can be used in SQL expression.
    */
-  public void registerUdf(String functionName, Class<? extends BeamSqlUdf> clazz) {
-    schema.add(functionName, ScalarFunctionImpl.create(clazz, BeamSqlUdf.UDF_METHOD));
+  public synchronized void registerUdf(String functionName, Class<? extends BeamSqlUdf> clazz) {
+    validateUdfNameUniqueness(functionName);
+    udfs.put(functionName, clazz);
   }
 
   /**
    * register {@link SerializableFunction} as a UDF function which can be used in SQL expression.
    * Note, {@link SerializableFunction} must have a constructor without arguments.
    */
-  public void registerUdf(String functionName, SerializableFunction sfn) {
-    schema.add(functionName, ScalarFunctionImpl.create(sfn.getClass(), "apply"));
+  public synchronized void registerUdf(String functionName, SerializableFunction sfn) {
+    validateUdfNameUniqueness(functionName);
+    sfnUdfs.put(functionName, sfn);
   }
 
   /**
    * Register a UDAF function which can be used in GROUP-BY expression.
    * See {@link BeamSqlUdaf} on how to implement a UDAF.
    */
-  public void registerUdaf(String functionName, Class<? extends BeamSqlUdaf> clazz) {
-    schema.add(functionName, AggregateFunctionImpl.create(clazz));
+  public synchronized void registerUdaf(String functionName, Class<? extends BeamSqlUdaf> clazz) {
+    validateUdfNameUniqueness(functionName);
+    udafs.put(functionName, clazz);
   }
 
-  /**
-   * Registers a {@link BaseBeamTable} which can be used for all subsequent queries.
-   *
-   */
-  public void registerTable(String tableName, BaseBeamTable table) {
-    schema.add(tableName, new BeamCalciteTable(table.getRowType()));
-    planner.getSourceTables().put(tableName, table);
+  private void validateUdfNameUniqueness(String name) {
+    if (udfs.containsKey(name) || udafs.containsKey(name) || sfnUdfs.containsKey(name)) {
+      throw new IllegalArgumentException("UDF named: " + name + " is already registered!");
```

Comment:
Do we really want this check? I thought we were planning on allowing overrides to registrations. This will be particularly useful once we have service loader registration. You might automatically register a version of a UDF/UDAF that you don't actually want to use, in which case we want to provide a way to override that.